test/libsolutil/LP.cpp

```diff
@@ -0,0 +1,314 @@
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolutil/LP.h>
+#include <libsolutil/LinearExpression.h>
+#include <libsolutil/CommonIO.h>
+#include <libsmtutil/Sorts.h>
+#include <libsolutil/StringUtils.h>
+#include <test/Common.h>
+
+#include <boost/test/unit_test.hpp>
+
+using namespace std;
+using namespace solidity::smtutil;
+using namespace solidity::util;
+
+
+namespace solidity::util::test
+{
+
+class LPTestFramework
+{
+public:
+	LPTestFramework()
+	{
+		m_solvingState.variableNames.emplace_back("");
+	}
+
+	LinearExpression constant(rational _value)
+	{
+		return LinearExpression::factorForVariable(0, _value);
+	}
+
+	LinearExpression variable(string const& _name)
+	{
+		return LinearExpression::factorForVariable(variableIndex(_name), 1);
+	}
+
+	/// Adds the constraint "_lhs <= _rhs".
+	void addLEConstraint(LinearExpression _lhs, LinearExpression _rhs)
+	{
+		_lhs -= _rhs;
+		_lhs[0] = -_lhs[0];
+		m_solvingState.constraints.push_back({move(_lhs), false});
+	}
+
+	void addLEConstraint(LinearExpression _lhs, rational _rhs)
+	{
+		addLEConstraint(move(_lhs), constant(_rhs));
+	}
+
+	/// Adds the constraint "_lhs = _rhs".
+	void addEQConstraint(LinearExpression _lhs, LinearExpression _rhs)
+	{
+		_lhs -= _rhs;
+		_lhs[0] = -_lhs[0];
+		m_solvingState.constraints.push_back({move(_lhs), true});
+	}
+
+	void addLowerBound(string _variable, rational _value)
+	{
+		size_t index = variableIndex(_variable);
+		if (index >= m_solvingState.bounds.size())
+			m_solvingState.bounds.resize(index + 1);
+		m_solvingState.bounds.at(index).lower = _value;
+	}
+
+	void addUpperBound(string _variable, rational _value)
+	{
+		size_t index = variableIndex(_variable);
+		if (index >= m_solvingState.bounds.size())
+			m_solvingState.bounds.resize(index + 1);
+		m_solvingState.bounds.at(index).upper = _value;
+	}
+
+	void feasible(vector<pair<string, rational>> const& _solution)
+	{
+		auto [result, model] = m_solver.check(m_solvingState);
+		BOOST_REQUIRE(result == LPResult::Feasible);
+		for (auto const& [var, value]: _solution)
+			BOOST_CHECK_MESSAGE(
+				value == model.at(var),
+				var + " = "s + ::toString(model.at(var)) + " (expected " + ::toString(value) + ")"
+			);
+	}
+
+	void infeasible()
+	{
+		auto [result, model] = m_solver.check(m_solvingState);
+		BOOST_CHECK(result == LPResult::Infeasible);
+	}
+
```

Comment:
```suggestion

        void sat()
        {
                auto [result, model] = m_solver.check(m_solvingState);
                BOOST_REQUIRE(result == LPResult::Feasible);
        }
        
```

Reply:
What about changing `feasible` above to make `_solution` optional?

Reply:
Actually I think it's better to always require a solution, then we can better check for any nondeterminism.